django/db/models/fields/__init__.py

```diff
@@ -144,6 +148,7 @@ def __init__(self, verbose_name=None, name=None, primary_key=False,
         self.remote_field = rel
         self.is_relation = self.remote_field is not None
         self.default = default
+        self.db_default = db_default
```

Comment:
Do we want to allow setting both `default` and `db_default`? Might make sense if one only wants to use `db_default` during a deploy to prevent downtime.

Reply:
I think it's an edge-case, but I don't see a strong reason to forbid it. I think `default` should take precedence over `db_default`.

Reply:
Only literal values of functions with literal values arguments should be allowed here. 

No _resolving_ phase should be necessary if an expression is provided, the only requirement should be an `as_sql` function that gets passed a `compiler` and a `connection` through `compiler.compile`.

The way it's currently defined is too lax because it allows field references and event JOINs

Reply:
I'm not sure what you're suggesting here.

Reply:
What I mean is that it should not be allowed to pass expressions of the form `Lower('foo')` (reference to a field) but your patch currently allows that and has inherent complexity because it does (e.g. `resolve_expressions`).

Only expressions that resolve to SQL literals or functions with all of the arguments resolving to SQL literals (on backends that support it) should be allowed here. Feels like we need a backend hook similar to `supports_expression` (e.g. `supports_field_default_expression`) that could be called by a check and called again on compilation attempts.